README.md

```diff
@@ -55,8 +52,17 @@ Express constraints and react accordingly, as an option to lenient handling.
                   .takingAtLeast(10)
                   .takingAtMost(20)
                   .waitingFor(2, ChronoUnit.SECONDS)
-                  .orThrow(TooFewOffersConstraintException::new)
-                  .collect(Collectors.toList());
+                  .orThrow(TooFewOffersConstraintException::new);
+```
+
+Optional capabilities, to handle exceptions and errors are built-in and very
+easy to use.
+
+```java
+  var offers = Queries.queryFor("offers/rental", NewOffer.class)
+                  .waitingFor(400)
+                  .onError(error -> LOG.error("Failed!", error))
```

Comment:
So `onError()` is only a posibility to do side effects in case of an error, but the query always evaluate to an result (if not ended with `orThrows()`. Am I right?

Reply:
Yes, this example tries to show that.

Reply:
I like your comment above though, perhaps the 3 ending-states imply the API terminology; leaning towards names with `error`, `success` (or the inverse of it) and `timeout`.

When we were reviewing this you hinted at `onMissing` or `onEmpty`, how would that look?

Reply:
Yeah, would be nice to have possibilities to side effect actions for all of the tree ending-stages. Maybe `onSuccess`, `onTimeout`, `onError`?
Ending operations could be: `orEmpty`, `orDefaults` (this are alternatives to success)

Reply:
Those are great ideas. Was your thought with the side-effect actions to provide a way of composing optional or chaining queries? I would rather see that as not so desirable.

One direct technical question would be, in the `on`-actions, do we still block on the calling thread?

I'm leaning towards having a smaller API now, and trying to move towards the best possible consistency towards the user. Things like providing the best possible micro-timing for the blocking call, to be tuned to the expected query-timeout (not counting GC).

So:

`onErrorReturn` - is out and, like you said, not useful.

`onError` - is useful, as the initial smallest possible addition that will bring transparency to developers, that want to introspect and have side-effects on errors.

**The query always blocks, and always returns or throws.**

That API is `orEmpty` or `orDefaults` or optional `orThrows`.

The stream is out and queries return `Collection<T>`. Users can move to stream and do filter, map and use collectors as they feel necessary.